main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala

```diff
@@ -0,0 +1,650 @@
+      val (primEnums, enumsWithObjField) = li.partition(_._2._2.isPrimitive)
+      sym -> (primEnums.map(x => (x._2._1, x._2._2)).toSet, enumsWithObjField.map(x => x._2._1).toSet)
+    }.toMap // Despite IDE highlighting, this is actually necessary.
+
+    // 5. Generate bytecode for each enum case.
+    val enumByteCodes = enumsGroupedBySymbolAndFieldType.map{ case (sym, (primEnums, objectEnums)) =>
+      val generatedPrimEnums : Map[(String, Type), Array[Byte]] = primEnums.map{ case (name, tpe) =>
+        val qualName = CodegenHelper.getEnumCaseName(sym, name, Some(tpe))
+        val interface = enumInterfaces(sym)._1
+        (name, tpe) -> EnumGen.compileEnumClass(qualName, interface, Some(tpe), interfaceNames, flix.javaVersion)
+      }.toMap
+      val generatedObjectEnums : Map[String, Array[Byte]] = objectEnums.map{ name =>
+        val fullName = CodegenHelper.getEnumCaseName(sym, name, None)
+        val interface = enumInterfaces(sym)._1
+        name -> EnumGen.compileEnumClass(fullName, interface, None, interfaceNames, flix.javaVersion)
+      }.toMap
+      sym -> (generatedPrimEnums, generatedObjectEnums)
+    }
+
+    val e = System.nanoTime() - t
+    root.copy(time = root.time.copy(enumGen = e), enumInterfaceByteCodes = enumInterfaces, enumClassByteCodes = enumByteCodes).toSuccess
+  }
+
+  /**
+    * Generates an interface for each enum.
+    * Each case of the enum implements this interface. This interface extends `TagInterface`.
+    * For example, for enum `Result` we create:
+    *
+    * package ca.waterloo.flix.enums.Result;
+    * import ca.uwaterloo.flix.api.TagInterface;
+    * public interface EnumInterface extends TagInterface {
+    * }
+    *
+    * @param qualName Qualified Name of the interface to be generated
+    * @return byte code representation of the class
+    */
+  def compileEnumInterface(qualName: QualName)(flix: Flix): Array[Byte] = {
+    val visitor = new ClassWriter(0)
+
+    visitor.visit(flix.javaVersion, ACC_PUBLIC + ACC_ABSTRACT + ACC_INTERFACE, decorate(qualName), null,
+      asm.Type.getInternalName(Constants.objectClass), Array(asm.Type.getInternalName(Constants.tagInterface)))
+
+    visitor.visitEnd()
+    visitor.toByteArray
+  }
+
+  /**
+    * Generates the class for each enum case.
+    * Each class implements the interface of enum.
+    * @param qualName Qualified name of the generated class
+    * @param interfaceName Qualfied name of the interface of the Enum
+    * @param fType Type of the `value` field, None means that the type is not a primitive
+    * @param interfaces Map of functional interfaces to their qualified name
+    * @return byte code representation of the class
+    */
+  def compileEnumClass(qualName: QualName,
+                       interfaceName: QualName,
+                       fType: Option[Type],
+                       interfaces: Map[Type, QualName],
```

Comment:
This map should not really be necessary here. Instead it should be computed when needed.

We can leave this as as separate refactoring.